core/src/saros/net/stream/Socks5StreamService.java

```diff
@@ -99,714 +56,217 @@
-    log.debug(prefix() + "canceling response connection as it is not needed");
-
-    Thread waitToCloseResponse =
-        new Thread("CloseUnneededResponseConnection") {
-
-          @Override
-          public void run() {
-            try {
-              closeQuietly(future.get());
-            } catch (InterruptedException e) {
-              // nothing to do here
-            } catch (ExecutionException e) {
-              log.debug(
-                  prefix()
-                      + "Exception while waiting to close unneeded connection: "
-                      + e.getMessage());
-            }
-          }
-        };
-    waitToCloseResponse.start();
-  }
-
-  private String verboseLocalProxyInfo() {
-    final Socks5Proxy proxy = NetworkingUtils.getSocks5ProxySafe();
-    final StringBuilder builder = new StringBuilder();
-
-    builder.append("local SOCKS5 proxy");
-
-    if (!proxy.isRunning()) {
-      builder.append(" not running");
-      return builder.toString();
-    }
-
-    builder.append(" running [");
-    builder.append("port=").append(proxy.getPort()).append(",");
-    builder.append(" configured streamhosts=");
-    builder.append(proxy.getLocalAddresses());
-    builder.append("]");
-    return builder.toString();
-  }
-
-  /*
-   * the Smack API does not tell us, if a mediated Socks5 connection is
-   * unidirectional. But some server implementation (OpenFire) may return such
-   * a connection. In this case we have to wrap the unidirectional
-   * connections.
-   */
+  // *********** IStreamService impl start
+  @Override
+  public IByteStreamConnection connect(final String connectionId, final JID remoteAddress)
+      throws IOException, InterruptedException {
 
-  /**
-   * Tests one of the bytestreams != null in the opposite direction. It returns it if bidirectional
-   * or tries to wrap two unidirectional streams if possible. Else an exception is thrown. The
-   * testing order is defined by the boolean preferInSession.
-   *
-   * @pre inSession!=null || outSession!=null
-   * @param inSession
-   * @param outSession
-   * @param preferInSession which stream to test preferable (if != null)
-   * @return a bidirectional BytestreamSession
-   * @throws IOException if there is only one unidirectional session
-   */
-  private BytestreamSession testAndGetMediatedBidirectionalBytestream(
-      Socks5BytestreamSession inSession,
-      Socks5BytestreamSession outSession,
-      boolean preferInSession)
-      throws IOException {
+    Objects.requireNonNull(connectionId, "connectionId must not be null");
+    Objects.requireNonNull(remoteAddress, "remoteAddress must not be null");
 
-    String msg = prefix() + "response connection is mediated, too, ";
+    if (connectionId.isEmpty())
+      throw new IllegalArgumentException("connectionId must not be empty");
 
-    Socks5BytestreamSession session = preferInSession ? inSession : outSession;
+    log.debug(
```

Comment:
```suggestion
    log.debug("establishing Socks5 bytestream to: {}, {}...", () -> remoteAddress, () -> verboseLocalProxyInfo())
```

Reply:
Guess that this is my opinion but I find it very frustrating that you simply cannot mix up lambdas and objects.

Reply:
I had the same expectation and was very confused about the error messages when I tried to mix both. However, I think it is still better than log4j-1.2.